src/validation/func.rs

```diff
@@ -610,47 +608,50 @@ impl<'a> FunctionValidationContext<'a> {
 
-	fn pop_value(&mut self, value_type: StackValueType) -> Result<(), Error> {
-		self.check_stack_access()?;
-		match self.value_stack.pop()? {
-			StackValueType::Specific(stack_value_type) if stack_value_type == value_type => Ok(()),
-			StackValueType::Any => Ok(()),
-			StackValueType::AnyUnlimited => {
-				self.value_stack.push(StackValueType::AnyUnlimited)?;
-				Ok(())
-			},
-			stack_value_type @ _ => Err(Error(format!("Expected value of type {:?} on top of stack. Got {:?}", value_type, stack_value_type))),
-		}
-	}
-
-	fn tee_value(&mut self, value_type: StackValueType) -> Result<(), Error> {
-		self.check_stack_access()?;
-		match *self.value_stack.top()? {
-			StackValueType::Specific(stack_value_type) if stack_value_type == value_type => Ok(()),
-			StackValueType::Any | StackValueType::AnyUnlimited => Ok(()),
-			stack_value_type @ _ => Err(Error(format!("Expected value of type {:?} on top of stack. Got {:?}", value_type, stack_value_type))),
+	fn pop_value(&mut self, value_type: StackValueType) -> Result<StackValueType, Error> {
+		let (is_stack_polymorphic, label_value_stack_len) = {
+			let frame = self.top_label()?;
+			(frame.polymorphic_stack, frame.value_stack_len)
+		};
+		let stack_is_empty = self.value_stack.len() == label_value_stack_len;
+		let actual_value = if stack_is_empty && is_stack_polymorphic {
+			StackValueType::Any
+		} else {
+			self.check_stack_access()?;
+			self.value_stack.pop()?
+		};
+		match actual_value {
+			StackValueType::Specific(stack_value_type) if stack_value_type == value_type => {
+				Ok(actual_value)
+			}
+			StackValueType::Any => Ok(actual_value),
+			stack_value_type @ _ => Err(Error(format!(
+				"Expected value of type {:?} on top of stack. Got {:?}",
+				value_type, stack_value_type
+			))),
 		}
 	}
 
-	fn pop_any_value(&mut self) -> Result<StackValueType, Error> {
-		self.check_stack_access()?;
-		match self.value_stack.pop()? {
-			StackValueType::Specific(stack_value_type) => Ok(StackValueType::Specific(stack_value_type)),
-			StackValueType::Any => Ok(StackValueType::Any),
-			StackValueType::AnyUnlimited => {
-				self.value_stack.push(StackValueType::AnyUnlimited)?;
-				Ok(StackValueType::Any)
-			},
+	fn check_stack_access(&self) -> Result<(), Error> {
+		let value_stack_min = self.frame_stack.top().expect("at least 1 topmost block").value_stack_len;
+		if self.value_stack.len() > value_stack_min {
+			Ok(())
+		} else {
+			Err(Error("Trying to access parent frame stack values.".into()))
 		}
 	}
 
-	fn tee_any_value(&mut self) -> Result<StackValueType, Error> {
-		self.check_stack_access()?;
-		Ok(self.value_stack.top().map(Clone::clone)?)
+	fn tee_value(&mut self, value_type: StackValueType) -> Result<StackValueType, Error> {
+		let value = self.pop_value(value_type)?;
+		self.push_value(value)?;
+		Ok(value)
 	}
 
 	fn unreachable(&mut self) -> Result<(), Error> {
-		Ok(self.value_stack.push(StackValueType::AnyUnlimited)?)
+		let frame = self.frame_stack.top_mut()?;
```

Comment:
Not sure I agree with this change

Much cleaner to push something to stack than to mutate the last entry in obscure fasion

Is there really no way to preserve former logic?

Reply:
And I'm not sure about having some special value with unclear behavior to signal that stack is in polymorphic state is much better :-) 

What's bad in taking current frame and marking it is transitioned to polymorphic stack state?